src/language-js/print/function-parameters.js

```diff
@@ -223,6 +225,36 @@ function shouldGroupFunctionParameters(functionNode, returnTypeDoc) {
   );
 }
 
+/**
+ * The "decorated function" pattern.
+ * The arrow function should be kept hugged even if its signature breaks.
+ *
+ * ```
+ * const decoratedFn = decorator(param1, param2)((
+ *   ...
+ * ) => {
+ *   ...
+ * });
+ * ```
+ * @param {AstPath} path
+ */
+function isDecoratedFunction(path) {
+  return path.match(
+    (node) =>
+      node.type === "ArrowFunctionExpression" &&
+      node.body.type === "BlockStatement",
+    (node, name, number) =>
+      node.type === "CallExpression" &&
+      name === "arguments" &&
+      number === 0 &&
+      node.arguments.length === 1 &&
+      node.callee.type === "CallExpression" &&
+      node.callee.callee.type === "Identifier",
+    (node, name) => node.type === "VariableDeclarator" && name === "init",
+    (node) => node.kind === "const"
```

Comment:
Why it has to be `const`, some people don't like `const` https://www.npmjs.com/package/eslint-plugin-prefer-let

Reply:
I believe these people are in minority. I wanted this pattern detection to be as specific as possible to exclude false positives.

Reply:
I don't like it, and what about people don't follow `one-var`? He can't use `const` because `foo` will reassign

**Prettier pr-10714**
[Playground link](https://deploy-preview-10714--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuc0DOMAEBhCBXWOAJ0wF5MATOSIgQxgiIAoAdEAWwE8BaSA+ImwCUrKJkxMADkQiS0STMEwBLKMpjLaAG1z8A-Aqh52AI2IBuTFtpmtBzBiKqA5pgC+QsgD5FLMeMxIKAxMSTJMPDQ4ABE4ADNaPC0YKRk5ABpffwCVNQ1tXVgFAAZ0vxzxa1sFNkKBNjLs9yFzP3KAoJCAbTRMqJgAXXDIuABlGHo4JiU+ItCAOlV1TR18WGbWps6saGwtZTAAa3CmT1IfftZUfgaJdTOfdUwAakwARha2pqI4GDwiMSnbwSdo5AA8JjwMAYYmgYH2R1IwF2CMObi8oIqwEk8yqcC0bgUwDQ81mMDcmPEYIA9JDodAMU1Pv4KVAhJk4hAIOE3ps-Fpfjg1gJwlQaPRGFcuLxhcRhKJxKlZPJFLllgVhfYjKYLFYbPj7I4XM1gcBMdtQsMorEEkkUtJlZkzU1xEt8qt+CVGhU9dVMLVZYIQN6Ah4+Vt0Fgen1fkNyCNxpNpoFhQocW6VnUNubI5gUQdjuQgecHL8rmTbkx7sCnq8PuGcj8-gCJA8QS7MBCoTC81B4QWkfmjujKQFsbj9QSiSSyayfTS6TDGTlmeJWezMJzueRecGQLINOhkKBaEQZAB3AAKp4QaGQIG059onDv6RAJjoR1+o0ktDALmQGAiDwOA3zgHUKCoCgABlaCgZw8FoZw4AAMUYdh6A0eD70SBg9wACxgdgtAAdXw9Q4DQX8wDGW9lgAN3UTh7zANBXxAVQoiIGBLzoZwMOQBItCiN8ACs0AADwAIU-Q5v1odg4Gg1Q4EE7QRJAcSJNGFwBQARTwCB4DU4TQJAX8iC4+8TEnPdpFUGASOUCgYHw5AAA5SnMmQohIuhJHvaRKOIejVLfABHQz4F45UcLQbgoDgOAoL3H5IuUH5eKQgSkCEjSonYZRAOAsy0F0uADKM1TcvUsyJhMJyXLcpAACY3yA2hlH2eDcHYHKQEogBWPcRgAFRsO8atMt96JAgBJKAqFgUYwCcSQYAAQUW8ZOAFEyojcNwgA)
<!-- prettier-ignore -->
```sh
--parser babel
```

**Input:**
<!-- prettier-ignore -->
```jsx
const Counter = decorator("my-counter")(
  (props: { initialCount?: number; label?: string }) => {
    const p = useDefault(props, {
      initialCount: 0,
      label: "Counter",
    });

    const [s, set] = useState({ count: p.initialCount });
    const onClick = () => set("count", (it) => it + 1);

    return () => (
      <button onclick={onClick}>
        {p.label}: {s.count}
      </button>
    );
  }
), foo = 1;

let Counter = decorator("my-counter")(
  (props: { initialCount?: number; label?: string }) => {
    const p = useDefault(props, {
      initialCount: 0,
      label: "Counter",
    });

    const [s, set] = useState({ count: p.initialCount });
    const onClick = () => set("count", (it) => it + 1);

    return () => (
      <button onclick={onClick}>
        {p.label}: {s.count}
      </button>
    );
  }
), foo = 1;
```

**Output:**
<!-- prettier-ignore -->
```jsx
const Counter = decorator("my-counter")((props: {
    initialCount?: number,
    label?: string,
  }) => {
    const p = useDefault(props, {
      initialCount: 0,
      label: "Counter",
    });

    const [s, set] = useState({ count: p.initialCount });
    const onClick = () => set("count", (it) => it + 1);

    return () => (
      <button onclick={onClick}>
        {p.label}: {s.count}
      </button>
    );
  }),
  foo = 1;

let Counter = decorator("my-counter")(
    (props: { initialCount?: number, label?: string }) => {
      const p = useDefault(props, {
        initialCount: 0,
        label: "Counter",
      });

      const [s, set] = useState({ count: p.initialCount });
      const onClick = () => set("count", (it) => it + 1);

      return () => (
        <button onclick={onClick}>
          {p.label}: {s.count}
        </button>
      );
    }
  ),
  foo = 1;

```

Reply:
If there are multiple variables in one declaration, another level of indentation is added. That means the author explicitly doesn't care about the indentatation in the function body, so that would be a false positive. If the language had function decorators, decorated functions would be function declarations (most probably top-level ones), not function expressions, and function declarations are normally not reassigned (even though the language allows that):

```js
@component('my-counter')
function Counter({ initialCount = 0, label = 'Counter' }: CounterProps) {
  // ...
});
```

But the language doesn't support this (so far?), so the natural thing to do instead is to use `const`.

Reply:
> You didn't check `declarations` length

I'll add this.